File: 프로그래머스/lv2/84512. 모음 사전/모음 사전.js
function solution(word) {
    const alphabet = ['A', 'E', 'I', 'O', 'U'];

    const dfs = (cur, cnt, complete) => {
        cur += alphabet[0];

        if (cur.length > alphabet.length) {
            cur = cur.slice(0, cur.length - 1).split('');

            for (let i = cur.length - 1; i >= 0; --i) {
                if (complete > i)
                    break;

                if (i === complete && cur[complete] === alphabet[alphabet.length - 1]) {
                    ++complete;
                    if (i + 1 < cur.length) cur[i + 1] = alphabet[alphabet.length - 1];
                    break;
                }

                if ((alphabet.indexOf(cur[i]) + 1) % alphabet.length === 0) {
                    cur[i] = 'z';
                }
                else {
                    cur[i] = alphabet[(alphabet.indexOf(cur[i]) + 1)];
                    break;
                };
            }
            cur = cur.filter((e) => e !== 'z').join().replaceAll(',', '');
        }

        if (cur === word)
            return cnt + 1;

        return dfs(cur, cnt + 1, complete);
    };

    return dfs('', 0, 0);
}

File: 프로그래머스/lv2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int solution(string word) {
    int answer = 0;
    vector<char> v = { 'A', 'E', 'I', 'O', 'U' };

    for (int i = 0; i < word.length(); ++i)
    {
        int cnt = 0, sum = 0;
        while (cnt < 5 - i)
            sum += pow(5, cnt++);

        auto idx = find(v.begin(), v.end(), word[i]) - v.begin();
        answer += sum * idx + 1;
    }

    return answer;
}